add obi troubleshooting docs (#8559)
open-telemetry/opentelemetry.io: content/en/docs/zero-code/obi/troubleshooting.md

---
title: Troubleshooting
description: Troubleshooting OBI common issues and errors
weight: 22
cSpell:ignore: Clickhouse
---

On this page, you can learn how to diagnose and resolve common OBI errors and
issues.

## Troubleshooting tools

OBI provides a variety of tools and configuration options to help diagnose and
troubleshoot issues.

> ### Detailed Logging

Reviewer: ```suggestion
### Detailed logging
```

You can increase the logging verbosity of OBI by setting the `log_level`
configuration or the `OTEL_EBPF_LOG_LEVEL` environment variable to `debug`. This
will provide more detailed logs that may help in diagnosing issues.

To enable logging from the BPF programs, set the `ebpf.bpf_debug` configuration
or the `OTEL_EBPF_BPF_DEBUG` environment variable to `true`. **Use this only for
debugging**, as it can generate a significant number of logs.

### Configuration Logging

> By default, OBI merges its configuration from three different sources, from less

Reviewer: ```suggestion
By default, OBI merges its configuration from three different sources, from least
```

to most priority:

- Built-in default configuration
- Configuration file (provided via the `--config` flag or
  `OTEL_EBPF_CONFIG_PATH`)
- Environment variables (usually starting with `OTEL_EBPF_`)

It is often helpful to view the final merged configuration. Using the
`log_config` configuration value (or `OTEL_EBPF_LOG_CONFIG` environment
variable), you can instruct OBI to log the final configuration at startup.

`log_config` supports the following values:

- `yaml` — logs the final configuration in YAML format; best for human
  readability since it matches the config file structure
- `json` — logs the final configuration in JSON format; best for log shippers
  since it is a single structured line

> ### Internal Metrics

Reviewer: ```suggestion
### Internal metrics
```

     -> service.namespace: Str(opentelemetry-demo)
     -> k8s.owner.name: Str(flagd)
     -> k8s.kind: Str(Deployment)
     -> k8s.replicaset.name: Str(flagd-5cccb4c4f5)
     -> k8s.pod.name: Str(flagd-5cccb4c4f5-sfkcm)
     -> k8s.container.name: Str(flagd)
     -> k8s.deployment.name: Str(flagd)
     -> service.version: Str(2.0.2)
     -> k8s.namespace.name: Str(default)
     -> otel.library.name: Str(go.opentelemetry.io/obi)
ScopeSpans #0
ScopeSpans SchemaURL:
InstrumentationScope
Span #0
    Trace ID       : 63a2723a58e0033170e58b1ff27ef03d
    Parent ID      :
    ID             : fab47609b60cc4e0
    Name           : /opentelemetry.proto.collector.metrics.v1.MetricsService/Export
    Kind           : Client
    Start time     : 2025-11-28 16:10:35.4241749 +0000 UTC
    End time       : 2025-11-28 16:10:35.42555658 +0000 UTC
    Status code    : Unset
    Status message :
Attributes:
     -> rpc.method: Str(/opentelemetry.proto.collector.metrics.v1.MetricsService/Export)
     -> rpc.system: Str(grpc)
     -> rpc.grpc.status_code: Int(0)
     -> server.address: Str(otel-collector.default)
     -> peer.service: Str(otel-collector.default)
     -> server.port: Int(4317)
```

> ### Performance Profiler (pprof)

Reviewer: ```suggestion
### Performance profiler (pprof)
```

OBI can expose a `pprof` port to allow performance profiling. To enable it, set
the `profile_port` configuration value or the `OTEL_EBPF_PROFILE_PORT`
environment variable to the desired port.

This is an advanced use case and typically not required.

## Common OBI issues

This section covers how to resolve common OBI issues.

> ---

Reviewer: ```suggestion

```

### Node.js services crash or become unresponsive when OBI is running

To enable better context propagation in Node.js applications, OBI injects custom
code to track the current execution context. It does so using the Node.js
inspector protocol and sends the `SIGUSR1` signal to the Node process to open
the inspector.

However, if the application defines its own `SIGUSR1` signal handler, It will

> handle OBI's signal in a custom way, which may cause crashes or unresponsiveness

Reviewer: ```suggestion
However, if the application defines its own `SIGUSR1` signal handler, it
handles OBI's signal in a custom way, which may cause crashes or unresponsiveness
```

of the targeted application. For example:

```javascript
process.on('SIGUSR1', () => {
  process.exit(0);
});
```

Or by using Node.js flags that register their own signal handling, such as:

```commandline
node --heapsnapshot-signal=SIGUSR1
```

**Solutions:**

- Use the `discovery` configuration to exclude specific Node.js applications
  from OBI tracking, preventing OBI from sending `SIGUSR1`.

> - Or disable Node.js context propagation entirely by setting

Reviewer: ```suggestion
- Disable Node.js context propagation entirely by setting
```
